### ISSUE TITLE
Using a `DynamoDBClient` instance with empty configuration result in error "config.endpoint is not a function"

### ISSUE THREAD

**Issue by danillouz** [aws/aws-sdk-js-v3#4811]
### Checkboxes for prior research

- [X] I've gone through [Developer Guide](https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide) and [API reference](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest)
- [X] I've checked [AWS Forums](https://forums.aws.amazon.com) and [StackOverflow](https://stackoverflow.com/questions/tagged/aws-sdk-js).
- [X] I've searched for [previous similar issues](https://github.com/aws/aws-sdk-js-v3/issues) and didn't find any solution.

### Describe the bug

After creating a DynamoDB client with:
- An empty config
- A config containing an `undefined` field, like `endpoint`

Creating/sending commands to the client will always result in the error:

```
config.endpoint is not a function
```

This used to work, but breaks starting with `3.195.0` and above, including the latest version `3.348.0`.
The last working version is `3.194.0`.

### SDK version number

@aws-sdk/client-dynamodb@3.195.0 up to @aws-sdk/client-dynamodb@3.348.0, @aws-sdk/lib-dynamodb@3.195.0 up to @aws-sdk/lib-dynamodb@3.348.0

### Which JavaScript Runtime is this issue in?

Node.js

### Details of the browser/Node.js/ReactNative version

v18.16

### Reproduction Steps

```ts
import { DynamoDBClient } from "@aws-sdk/client-dynamodb"
import { DynamoDBDocumentClient, GetCommand } from "@aws-sdk/lib-dynamodb"

async function main() {
  const client = new DynamoDBClient({})
  const docClient = DynamoDBDocumentClient.from(client)

  return docClient.send(
    new GetCommand({
      TableName: "mytable",
      Key: { PK: `mypk`, SK: `mysk` },
    })
  )
}

main().catch(err => {
  console.error(err) // Error: config.endpoint is not a function
})
```

Our specific use case is that the DynamoDB client config contains an `undefined` value for the field `endpoint` (because we only set it locally for testing):

```ts
import Settings from "@settings"

const client = new DynamoDBClient({
  endpoint: Settings.dynamoDbEndpoint, // Will be `undefined`
})
```

But both cases result in the same mentioned error.

### Observed Behavior

Fails with:

```
Error: config.endpoint is not a function
```

### Expected Behavior

I expected it to create/send commands to the client without failing.

### Additional Information/Context

I might have missed it, but I can't find anywhere (in the SDK docs or release notes) that the previous behavior has changed in newer versions.

Most relevant related issue (closed due to inactivity):
- https://github.com/aws/aws-sdk-js-v3/issues/4675

Related issues:
- https://github.com/aws/aws-sdk-js-v3/issues/4122
- https://github.com/aws/aws-sdk-js-v3/issues/4105
- https://github.com/aws/aws-sdk-js-v3/issues/4076
- https://github.com/aws/aws-sdk-js-v3/issues/3120

**Comment by RanVaknin** [comment]
Hi @danillouz ,

How are you setting your config? The issues you linked are all closed for a reason. in the version you mentioned there's a change in behavior in the endpoint resolution logic of all the newer SDKs. 

>Due to a recent change in AWS SDKs, AWS service endpoints are now resolved by an internal ruleset, found in client-*/endpoint/ruleset.ts. Example: https://github.com/kuhe/aws-sdk-js-v3/blob/1437c0f/clients/client-s3/src/endpoint/ruleset.ts

>Because of the potential complexity of these rulesets, a service endpoint can no longer be resolved without additional pieces of information from the various Command and Config objects.

In other words, you have to supply a region somehow. Whether you are using a shared credentials file in ~./aws/credentials a config file should be added, or via env variables, or explicitly in the client constructor. Then the SDK can resolve the routing rules based on the region provided. 

This is intended behavior.

**Comment by danillouz** [comment]
Hi @RanVaknin, thank you for looking into this!

We set the region via env var on the EC2 container, as documented [here](https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide/setting-region.html#setting-region-environment-variable). And we verified the value is set:

![image](https://github.com/aws/aws-sdk-js-v3/assets/6201287/e5ecc11b-2664-477b-b48a-df6f52561a53)

From what I understand from your comment, this should be sufficient for the SDK to resolve the routing rules?

**Comment by RanVaknin** [comment]
Hi @danillouz ,

Thanks for the additional context. 

I took a closer look and using your example, I was not able to reproduce this.

Considering this code:
```js
import { DynamoDB } from '@aws-sdk/client-dynamodb';
import { DynamoDBDocument } from '@aws-sdk/lib-dynamodb';

const client = new DynamoDB({ 
    endpoint: undefined 
});
const ddbDocClient = DynamoDBDocument.from(client);

try {
  const params = {
    // some parameters
  };

  const response = await ddbDocClient.scan(params);
  console.log('Response:', response);
} catch (err) {
  console.error(err);
}

// successful execution
```
With successful execution. 

Since the error message you are seeing mentions a function, I converted endpoint to return undefined but hitting the expected error.
```js
import { DynamoDB } from '@aws-sdk/client-dynamodb';
import { DynamoDBDocument } from '@aws-sdk/lib-dynamodb';

const client = new DynamoDB({ 
    endpoint: () => {return undefined} 
});
const ddbDocClient = DynamoDBDocument.from(client);

try {
  const params = {
    // some parameters
  };

  const response = await ddbDocClient.scan(params);
  console.log('Response:', response);
} catch (err) {
  console.error(err);
}

// "TypeError: Cannot destructure property 'hostname' of 'url' as it is undefined.  "
// expected output
```
____

Put the change in behavior aside, it sounds like all you need to do is null check before assigning the endpoint to account for test environments:

```js
const config = Settings.dynamoDbEndpoint ? { endpoint: Settings.dynamoDbEndpoint } : {};
const client = new DynamoDB(config); 
const ddbDocClient = DynamoDBDocument.from(client);
```

I hope this helps.

Thanks
Ran~

**Comment by danillouz** [comment]
Hi @RanVaknin, thank you for investigating!

Interesting, that it works for you..
I tried your suggestion before submitting the issue (i.e. using the ternary to pass `{}`), but that resulted in the same error for me.

So the way I understand it:
- There has been a change in behavior in the endpoint resolution logic of all the newer SDKs 
- But passing an empty `{}` config (or one with field `endpoint` set to `undefined`) should work in the newer versions

If so, then it looks like something is somehow "conflicting" with this behavior in our setup (but only in the newer versions).

I'll try to dig into this a bit more soon, and will post back with any findings for future reference.

**Comment by kuhe** [comment]
Please open a new issue and link to this if creating a new reproduction.

**Comment by wederribas** [comment]
I've faced the same `config.endpoint is not a function` issue when deploying the DynamoDB client using the `@aws-sdk/client-dynamodb` package on version 3.398.0.

The issue was related to what is mentioned in this issue https://github.com/aws/aws-sdk-js-v3/issues/4122.

Long story short, in my particular case, it was caused by the NewRelic NodeJS agent instrumentation code not playing well with the Dynamo client SDK during the endpoint automatic resolution.

NewRelic already addressed the issue in this PR: https://github.com/newrelic/node-newrelic-aws-sdk/pull/161/files.

Hopefully, running the latest version of Dynamo SDK and the NewRelic NodeJS agent should fix it. Unfortunately, my temporary solution was to downgrade the Dynamo SDK version to 3.190.0 until I'm able to upgrade the NewRelic agent.

**Comment by danillouz** [comment]
@wederribas thanks for sharing!

We also use newrelic, and just checked that the issue we ran into described in this ticket happened with `"newrelic": "9.7.5"`.

It looks like the newrelic PR you linked to that addresses the issue was fixed in `5.0.2`?
https://github.com/newrelic/node-newrelic-aws-sdk/pull/162

So not sure if that was the cause for us. But another explanation why this is happening (for us) is using both v2 and v3 of the AWS SDK: we have old code using v2, and new code using v3.

For what it's worth, I finally was able to refactor the code to upgrade everything to v3 (`3.425.0`) of the AWS SDK this week, and the issue no longer happens (but with `"newrelic": "10.1.2"`).
